### DISCUSSION TITLE
How to instrument an abstract function inside a class ?

### DISCUSSION THREAD

**Discussion by meirtolpin11** [open-telemetry/opentelemetry-java-instrumentation#8395] [Q&A]
Hello everyone, 

I am trying to instrument the `newDocumentBuilder` function inside `javax.xml.parsers.DocumentBuilderFactory` - 
```
    /**
     * Creates a new instance of a {@link javax.xml.parsers.DocumentBuilder}
     * using the currently configured parameters.
     *
     * @return A new instance of a DocumentBuilder.
     *
     * @throws ParserConfigurationException if a DocumentBuilder
     *   cannot be created which satisfies the configuration requested.
     */

    public abstract DocumentBuilder newDocumentBuilder()
        throws ParserConfigurationException;
```

This is my code snippet - 
```
public class DocumentBuilderFactoryInstrumentation implements TypeInstrumentation {
    @Override
    public ElementMatcher<TypeDescription> typeMatcher() {
        return extendsClass(named("javax.xml.parsers.DocumentBuilderFactory"));
    }

    @Override
    public void transform(TypeTransformer typeTransformer) {

        // DocumentBuilderFactory.newDocumentBuilder()
        typeTransformer.applyAdviceToMethod(
                named("newDocumentBuilder").and(takesArguments(0)),
                this.getClass().getName() + "$NewDocumentBuilder"
        );
    }
```

But unfortunately it's not working and `$NewDocumentBuilder` is never called. 
What am I doing wrong? it's working with all the other function which are part of `DocumentBuilderFactory` (there are not abstract).
So I believe that the reason it's not working is because it's abstract, but I have no idea how to solve that. 

Thanks!

**Comment by mateuszrzeszutek** [comment]
Hey @meirtolpin11 ,

The fragment you posted looks correct; can you also post the advice class? You could replace the advice enter/exit methods with some `System.out.println()` statements in case you don't want to share your code (that's also a good way to test that).

Instrumenting abstract methods definitely works in general, [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/java-http-client/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/httpclient/HttpHeadersInstrumentation.java)`s a sample instrumentation that does that in the agent.

**Comment by laurit** [comment]
You aren't actually instrumenting the abstract method, as abstract methods don't have a body, but rather the `newDocumentBuilder` method in a class that extends the `DocumentBuilderFactory`. Most likely this doesn't work because agent by default excludes some classes from instrumentation. For example you probably want to instrument `newDocumentBuilder` in `com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderFactoryImpl` but agent by default excludes classes under `com.sun` in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/fe56784f27dc4ae99d1e3eb5333c45d6892e5dea/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ignore/GlobalIgnoredTypesConfigurer.java#L87 If that is the case you can use your own `IgnoredTypesConfigurer` to override the exclude.
